How to update your code. (#10626)
ethereum/solidity: docs/080-breaking-changes.rst

  Use the "compact AST" (``--ast-compact--json`` resp. ``AST``) as replacement.

* The old error reporter (``--old-reporter``) has been removed.

How to update your code

> =======================

Reviewer: Do you think it would be a good idea to say something about panics here?

One thing I noticed when fixing external tests was that unit tests that depend on `revert` returning a specific message on checked math errors now fail. That's not really a Solidity problem and more related to external support for the new codes but that's still something we might want to let people know about.

Author: Tooling often does not even display these messages because the eth rpc interface does not support it. It is a breaking change and listed above (I hope) but it's not really anything you can change your (Solidity) code to fix it...

> =======================

Reviewer: Don't we also want to mention some of the less likely problems, like removing `log` functions or adding parentheses over exponentiation if anyone depended on the old operator order?

Especially the latter will just silently start producing wrong results instead of failing.

- If you rely on wrapping arithmetic, surround each operation with ``unchecked { ... }``.
- Optional: If you use SafeMath or a similar library, change ``x.add(y)`` to ``x + y``, ``x.mul(y)`` to ``x * y`` etc.

> - Add ``pragma abicoder v1;`` if you want to stay with the old ABI coder.

Reviewer 2: ```suggestion
- Add ``pragma abicoder v1;`` if you want to stay with the old ABI coder.
- Optionally remove ``pragma experimental ABIEncoderV2`` or ``pragma abicoder v2`` since it is redundant.
```

Reviewer: I'd put the extra part in a separate bullet point.

Reviewer 2: Fixed the suggestion!

- Change ``byte`` to ``bytes1``.

> - Add intermediate explicit type conversions if required.

Author: Not sure if we should go into detail here....

Reviewer: It still works as a checklist even without the detail but if I were going over it to update my code, having it all listed here would be helpful.

Author: I mean the compiler will also generate an error, so I'm not sure such a list is helpful here, especially as it can also only list categories and nothing you can search for in your code.

Reviewer: ok. I thought more about listing examples. Categories are indeed less useful.

I wonder if having this whole section is a good idea in general because after looking at details above it I noticed that only the most common cases are listed. I can see how it could still be attractive to people who don't want to spend time reading all of it but this way we're just encouraging more people not to read it :)

- Combine ``c.f{gas: 10000}{value: 1}()`` to ``c.f{gas: 10000, value: 1}()``.

> - Change ``msg.sender.transfer(x)`` to ``payable(msg.sender).transfer(x)`` or use a stored variable of ``address payable`` type.

Reviewer: `stored variable` sounds a bit too close to `storage variable` in my opinion and could be confusing because of that:
```suggestion
- Change ``msg.sender.transfer(x)`` to ``payable(msg.sender).transfer(x)`` or use an intermediate variable of ``address payable`` type.
```

Author: I really meant 'stored variable'. The reason is that you often compare `msg.sender` with `owner`, and then you can just make `owner` payable.

Reviewer: Oh. Ok then. But it would work just as well in a local variable, right?

  Use the "compact AST" (``--ast-compact--json`` resp. ``AST``) as replacement.

* The old error reporter (``--old-reporter``) has been removed.

How to update your code
=======================

- If you rely on wrapping arithmetic, surround each operation with ``unchecked { ... }``.
- Optional: If you use SafeMath or a similar library, change ``x.add(y)`` to ``x + y``, ``x.mul(y)`` to ``x * y`` etc.
- Add ``pragma abicoder v1;`` if you want to stay with the old ABI coder.

> - Optionally remove ``pragma experimental ABIEncoderV2`` or ``pragma abicoder v2`` since it is redundant.

Reviewer 3: Should mention that actually `pragma experimental ABIEncoderV2` is invalid.

Reviewer 3: Do we have a conversion step in solidity-upgrade for this?

Reviewer 2: There are no conversion steps in `solidity-upgrade` for 0.8.0.

Reviewer 3: Didn't we somewhere say that "solidity-upgrade does a lot of the manual work"?

Reviewer 2: Well nobody implemented anything and this was never prioritized.

Reviewer 3: I mean we have such a line in the docs or the blog post :)

Reviewer 2: We removed that line, from the blog post for 0.8.0!

> - Change ``byte`` to ``bytes1``.

Reviewer 3: Can we have a conversion step in solidity-upgrade for this?

- Add intermediate explicit type conversions if required.

> - Combine ``c.f{gas: 10000}{value: 1}()`` to ``c.f{gas: 10000, value: 1}()``.

Reviewer 3: Do we have a conversion step in solidity-upgrade for this?

> - Change ``msg.sender.transfer(x)`` to ``payable(msg.sender).transfer(x)`` or use a stored variable of ``address payable`` type.

Reviewer 3: Can we have a conversion step in solidity-upgrade for this?

> - Change ``x**y**z`` to ``(x**y)**z``.

Reviewer 3: Do we have a conversion step in solidity-upgrade for this?
